### PR TITLE
feat(web): handle controlled behavior of component

### LINKED ISSUE #1091: onChange handler and value behave weird and not consistent
**Affected Projects**
React

**Library Version:** x.y.z
v3.0.0-rc.25

**Describe the bug**
`onChange` handlers only become active when the component has a `value` set.
While i don't understand why this limitation is in place, the real issue for me is that `onChange` does not return a value that I can use for `value`.
E.g. MultiList returns a single string upon change, expects an array and internally uses an object to keep the current value in `state`.
For the RangeInput I had similar issues.

**To Reproduce**
Take `MultiList` and add `value` from state, and pass a function to `onChange` that sets the value to state. Actually, you wouldn't even need to set it and log it to console instead.

**Expected behavior**
I can easily store the data from an `onChange` in my global store and pass it back to the component using `value`

### PR DESCRIPTION
**Pull request by ShahAnuj2610**
fixes #1091

### CODE REVIEW

**Comment by bietkul** [comment]
@ShahAnuj2610 You are missing range components.

File: packages/web/src/components/list/MultiDataList.js [hunk 1]
```diff
@@ -385,7 +385,8 @@ class MultiDataList extends Component {
 		if (value === undefined) {
 			this.setValue(currentValue);
 		} else if (onChange) {
-			onChange(currentValue);
+			this.setValue(currentValue);
```
Comment (bietkul) [inline packages/web/src/components/list/MultiDataList.js]: @ShahAnuj2610 Why did you add `setValue` here?

File: packages/web/src/components/list/MultiDropdownList.js [hunk 2]
```diff
@@ -417,7 +415,8 @@ class MultiDropdownList extends Component {
 		if (value === undefined) {
 			this.setValue(currentValue);
 		} else if (onChange) {
-			onChange(currentValue);
+			this.setValue(currentValue);
+			onChange(Object.keys(this.state.currentValue));
```
Comment (bietkul) [inline packages/web/src/components/list/MultiDropdownList.js]: The same issue is here.

File: packages/web/src/components/list/MultiList.js [hunk 3]
```diff
@@ -457,7 +457,8 @@ class MultiList extends Component {
 		if (value === undefined) {
 			this.setValue(currentValue);
 		} else if (onChange) {
-			onChange(currentValue);
+			this.setValue(currentValue);
+			onChange(Object.keys(this.state.currentValue));
```
Comment (bietkul) [inline packages/web/src/components/list/MultiList.js]: The same issue is here.

File: packages/web/src/components/range/MultiDropdownRange.js [hunk 4]
```diff
@@ -220,9 +238,12 @@ class MultiDropdownRange extends Component {
 		const { value, onChange } = this.props;
 
 		if (value === undefined) {
-			this.selectItem(items);
+			this.selectItem({ item: items });
 		} else if (onChange) {
-			onChange(items);
+			this.selectItem({
```
Comment (bietkul) [inline packages/web/src/components/range/MultiDropdownRange.js]: We don't need to create any side effect when controlled behavior is being used, Just pass the value to the `onChange`.

**Review by bietkul — CHANGES_REQUESTED** [review]
I think the current implementation is not correct.
When the user defines the `value` with `onChange` then it is the user's responsibility to update the value prop accordingly.
It depends on the user whether he wants to update value or not.
Currently, you're just supplying the value to `onChange` after triggering the query i.e using `setValue`.

**Comment by ShahAnuj2610** [comment]
> I think the current implementation is not correct.
> When the user defines the `value` with `onChange` then it is the user's responsibility to update the value prop accordingly.
> It depends on the user whether he wants to update value or not.
> Currently, you're just supplying the value to `onChange` after triggering the query i.e using `setValue`.

@bietkul refactoring done

**Review by jyash97 — CHANGES_REQUESTED** [review]
@ShahAnuj2610  Can we avoid using Controlled Component in Examples as they are directly linked to docs via Codesandbox Embed links.

**Comment by ShahAnuj2610** [comment]
> @ShahAnuj2610 Can we avoid using Controlled Component in Examples as they are directly linked to docs via Codesandbox Embed links.

Since there is a behavioral change in the `onChange` method, shouldn't we include an example, just so the user is aware?

**Comment by bietkul** [comment]
I agree we should create different examples for controlled components.
For example, we already have `DataSearchControlled` example for `DataSearch`.

**Comment by ShahAnuj2610** [comment]
> I agree we should create different examples for controlled components.
> For example, we already have `DataSearchControlled` example for `DataSearch`.

examples added for controlled components

File: packages/web/src/components/list/MultiDataList.js [hunk 5]
```diff
@@ -385,7 +385,11 @@ class MultiDataList extends Component {
 		if (value === undefined) {
 			this.setValue(currentValue);
 		} else if (onChange) {
-			onChange(currentValue);
+			const newValue = Object.assign([], this.props.value);
+			const currentValueIndex = newValue.indexOf(currentValue);
+			if (currentValueIndex > -1) newValue.splice(currentValueIndex, 1);
+			else newValue.push(currentValue);
```
Comment (bietkul) [inline packages/web/src/components/list/MultiDataList.js]: Put this whole logic in some util function, please avoid code duplication.

**Review by bietkul — CHANGES_REQUESTED** [review]

**Review by bietkul — APPROVED** [review]

**Pull request merged**